File: DiagonalTraversal.java
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Traverse the matrix diagonally, alternating between upward and downward diagonals,
// and store the elements in the result array following the diagonal order.

// Your code here along with comments explaining your approach

public class DiagonalTraversal {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        int[] result = new int[m*n];
        int idx = 0; // To track the index in the result array
        int i = 0; int j = 0;
        boolean direction = true; // direction: true for upward, false for downward

        while (idx < m*n) {
            result[idx] = mat[i][j];
            idx++;

            // If the direction is upward
            if (direction) {
                if (j == n - 1) {
                    i++;
                    direction = false;
                } else if (i == 0) {
                    j++;
                    direction = false;
                } else {
                    i--;
                    j++;
                }
            } else {
                // If the direction is downward
                if (i == m - 1) {
                    j++;
                    direction = true;
                } else if (j == 0) {
                    i++;
                    direction = true;
                } else {
                    i++;
                    j--;
                }
            }
        }
        return result;
    }
}

File: DiagonalTraversal.py
# The approach uses O(m * n) time complexity, where m and n are the matrix dimensions.  
# It traverses the matrix diagonally, alternating between upward and downward directions,  
# adjusting row and column indices based on boundary conditions to fill the result array.  
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if mat == None or len(mat) == 0:
            return []
        
        m = len(mat)
        n = len(mat[0])
        result = [0 for i in range(m * n)]
        index = 0  # to track position in result
        r = 0  # row
        c = 0  # column
        Dir = 1  # direction: 1 means up, -1 means down
        
        while index < m * n:
            result[index] = mat[r][c]
            index += 1
            
            if Dir == 1:
                if c == n - 1:
                    r += 1
                    Dir = -1
                elif r == 0:
                    c += 1
                    Dir = -1
                else:
                    r -= 1
                    c += 1
            elif Dir == -1:
                if r == m - 1:
                    c += 1
                    Dir = 1
                elif c == 0:
                    r += 1
                    Dir = 1
                else:
                    r += 1
                    c -= 1
        
        return result
